### ISSUE TITLE
Physics Server should have configurable thread model

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot-proposals#483]
**Describe the project you are working on:**
I am working on a voxel engine, which is also used by other people.

**Describe the problem or limitation you are having in your project:**
As described in https://github.com/Zylann/godot_voxel/issues/124
In a voxel terrain, many meshes are getting created dynamically at runtime. Creating mesh colliders is 10 times slower than creating visual meshes (mostly due to computing its BVH), and that takes a lot of time. Unlike `VisualServer` and `PhysicsServer2D`, `PhysicsServer` has no thread model option, so it's basically not thread-safe and runs on the main thread. So I currently have no choice but to create my shapes on the main thread, which limits how many I can process per frame before starting to stall the game.
Beyond my own case, this also has the broader issue of preventing scenes from being loaded in a thread, since creating shape resources they might contain would be unsafe as well.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
If I take the same design as `VisualServer` and `PhysicsServer2D`, here is how it would be:
- `Single-Unsafe` would be the option for running physics on the main thread, without thread safety (basically how it is now).
- `Single-Safe` would still run physics on the main thread, but with thread-safety in case a thread calls its functions
- `Multithreaded` would run physics on a separate thread and also be safe to be called from other threads.

This would be a first improvement since the physics server would have its own thread time to spend without all the other stuff done in the main thread.

However there is a big flaw in this design, which is important for my use case: when I create a large amount of complex `Shape` resources, it calls Bullet which in turn has to build its BVH for that shape. So a call to the server is unavoidable. Then it will dequeue calls to create them, and that can stall the simulation because those calls will be processed in the same thread that runs the simulation, *even if those shapes don't intervene in the simulation yet*. Which defeats the reason to run in a thread in the first place.

This problem might exist in other servers already, however that's mostly an assumption I'm making since I don't know their code in detail (especially with Vulkan rewrite). What I'm asking here is that creating resources involving computations which don't intervene in the frame logic should not slow it down. If you can think of a good way to do this, please share.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
That section might be repetitive but if we go for the same design as other servers, `PhysicsServer` would be wrapped into a `PhysicsServerWrapMT` class which pushes calls to its function into a ring buffer, and dequeues calls from that buffer from inside the physics thread (or the main thread in case of `Single-Safe`).

But again, I'm not sure what's the best solution for the resources case. Maybe those could have another ring buffer and be processed in yet another thread? Or, being processed inside the calling thread, which would solve my problem since I already do the call from a thread.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
I already had a workaround so far by doing everything on the main thread, it limits performance  dramatically.

**Is there a reason why this should be core and not an add-on in the asset library?:**
This is already core for other servers, the feature is just missing in `PhysicsServer`. But I want to also emphasize on the resource creation case, which is the original reason why I bumped into this.

**Comment by Classy11** [comment]
I think that this feature would be a huge addon to Godot. PhysicsServers are often used for better performance and multi-threaded access; actually allowing the 3D Physics Servers to run on separate threads would be the coup de grâce and would turn them into performance monsters. It would unlock tremendously more power in most machines with gaming CPUs that have multiple threads.
This would also be great for dedicated servers, it could multiply their physics performance by many times based on how many cores are available.

Here is another proposal that wants to make use of more CPU cores for 3D Bullet physics but without the use of separate PhysicsServers, so a different solution that may be more suitable for different 3D Physics cases. #1155 .

**Comment by Hromon** [comment]
Yep, i need it.

**Comment by Calinou** [comment]
@Hromon Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by smix8** [comment]
It is not only an issue with mesh colliders but also 3D gameplay code that requires updated raycast positions from the physics server and/or run in _physics_process() in general.

Very common 3D game features like a characters head lookat or inverse kinematic hand and feet placement make the physics server stall the visual server so much that framerates are constantly halfed or worse with multithreading enabled while they run fine in singlethread mode.

**Comment by m4nu3lf** [comment]
Because of this issue when using godot_voxel in my project with a big terrain I was having consistent stuttering. To fix it I had to patch Godot in such a way that convex polygon shapes are created asynchronously within the BulletPhysicsServer.
I also had to add a way to flush the pending changes.
It's not a clean solution but it will allow me to keep working on my project.

**Comment by Malkverbena** [comment]
Any Update about this proposal?

**Comment by Calinou** [comment]
> Any Update about this proposal?

For Godot 4.0, we no longer need to cater to Bullet since only GodotPhysics will be provided b y default.

However, we don't have an active physics maintainer anymore, so this feature will take a while to be implemented (and may not be done in time for 4.0).

**Comment by Zylann** [comment]
Here is an update for the problem I initially had:

I am still heavily bottlenecked by shapes having to be created on the main thread. I have a job system cooking many meshes really quickly but it ends up most of the time waiting for the main thread. The ideal (and minimal) solution I would need would be the ability to fully create shapes from my own threads, in a safe way. I don't even need physics to *simulate* on a different thread, yet. 

To achieve this, the generic API might be affected a little. It depends if we want multithreading to be enforced by the generic API, or if it should be left to implementers. IMO it may depend on the chosen physics engine. I believe Godot Physics could have that feature.

Here are more detailed points I wrote earlier on RC:

1. **Have create_shape_offline(all_data_necessary), or a way to support it in the API**. The shape create implementation must not interact with simulation state until added to a body, I think that's common sense. Like, really just allocate whatever data is needed, precalculate whatever data structure, based on arguments given at the same time (like faces, BVH may be done here). Currently, several calls are needed sometimes (for example create, followed by `set_faces`). This API is a bit problematic because it opens the door to cases where the shape became attached to a body in the meantime. In a multithreaded scenario that means plenty of avoidable locking. Those functions could stay, but we should also be able to send all data in the create call. Because...

2. **Have it run on the caller thread safely.** For meshes and convex hulls, logic described in 1) is the expensive part. To have better control, it should really be executable from the caller's thread, even if the physics engine *simulates* in its own thread. This will also help cases of loading a scene in parallel, as shape resources will simply be able to be built from that thread. At the end of shape creation, the function should return a RID. That probably means inserting it into a RidContainer of some sort. It could be protected with a simple mutex or lockless thing, this is fine because insertion should be really quick compared to the work done in 1).

3. **Document the behavior per physic engine.** If these things are not possible due to how the physics engine is implemented (for example the ReactPhysics library uses a custom allocator which doesnt work in multithreaded scenarios), then it must be documented because for these engine implementations, having thread-safety will mean to lock the entire physics for the full call instead, and can cause enough stalls to be a show-stopper.

**Comment by Malkverbena** [comment]
@Zylann I've seen your contributions, especially in the double precision support. Noce job! 
Should this proposal be added to Physics 3D TRACKER?
Camille is not the physics maintainer anymore?

**Comment by Calinou** [comment]
> Camille is not the physics maintainer anymore?

No, as he went on to work at Rockstar San Diego (and is no longer working on Godot as a result).

**Comment by Zylann** [comment]
>  If I understood correctly, when you talk about multithreaded physics, the physics will be calculated on a separate thread from the main thread. But physics is still single-threaded. 

I actually edited my proposal recently, check `How it should be done`.

**Comment by KnightNine** [comment]
Anyone know if there's a PR which implements 3D threaded physics for 3.X?

**Comment by Calinou** [comment]
> Anyone know if there's a PR which implements 3D threaded physics for 3.X?

Not that I know of. Most of the development focus has been on 4.x lately.

**Comment by warent** [comment]
Just for clarification, does World3D utilize threads, so each world may be in its own thread?

According to the [documentation](https://docs.godotengine.org/en/4.2/classes/class_world3d.html#class-world3d-property-direct-space-state)
> When using multi-threaded physics, access is limited to Node._physics_process in the main thread.

And then looking at the setting;
![image](https://github.com/user-attachments/assets/2ccaa2ce-48ef-475f-a529-18c79cfdbd76)

It's a bit ambiguous. If I had to guess, it sounds like every World3D still shares the same thread even with multi-threaded physics enabled, correct?

**Comment by smix8** [comment]
@warent 
A World2D/3D is nothing more than an ID number that bundles together other ID numbers. It does not do anything by itself.

How a server decides to process all those parts that are relevant to it is a matter of each individual server implementation. PhysicsServer, RenderingServer, NavigationServer, AudioServer ... they all do threading differently or not at all.

Commonly the servers that use threads do so at a far lower level to speed things up, e.g. each individual item as a group task. Most projects have only 1 World2D/3D in use so having the threading at the world level would have no benefit for those projects.
